### PR TITLE
Add a macro to opt into aggressive constprop

### PR DESCRIPTION
**Pull request by Keno**
Right now aggressive constprop is essentially tied to the inlining
threshold (or to function's name being `getproperty` or `setproperty!`
respectively), which can be both somewhat brittle if the inlining cost
changes and insufficient when you do really know that const prop
would be beneficial even if the function is not inlineable. This
adds a simple macro that can be used to manually annotate methods
to force aggressive constprop on them.

### CODE REVIEW

File: base/expr.jl [hunk 1]
```diff
@@ -242,6 +242,19 @@ macro pure(ex)
     esc(isa(ex, Expr) ? pushmeta!(ex, :pure) : ex)
 end
 
+"""
+    @aggressive_constprop ex
+    @aggressive_constprop(ex)
+
+`@aggressive_constprop` requests more aggressive interprocedural constant
+propagation for the annotated function. For function where the return type
```
Comment (timholy) [inline base/expr.jl:250]:
```suggestion
propagation for the annotated function. For a method where the return type
```

File: test/core.jl [hunk 2]
```diff
@@ -7508,3 +7508,15 @@ let array = Int[]
 end
 @test compare_union37557(Ref{Union{Int,Vector{Int}}}(1),
                          Ref{Union{Int,Vector{Int}}}(1))
+
+# @aggressive_constprop ======
+g_nonaggressive(y, x) = (for i = 1:100; end; Val{x}())
```
Comment (timholy) [inline test/core.jl:7513]: Rather than the suprious loop, would it be clearer to just mark these as `@noinline`?

**Review by timholy — COMMENTED** [review]
Like it!

File: test/core.jl (hunk 2 above)
Comment (Keno) [reply test/core.jl:7513]: I thought I had tried that and it didn't work, but I guess I messed up somehow. Will do that instead.

File: test/core.jl [hunk 3]
```diff
@@ -7508,3 +7508,15 @@ let array = Int[]
...
+@noinline g_nonaggressive(y, x) = Val{x}()
+@noinline @Base.aggressive_constprop g_aggressive(y, x) = Val{x}()
+
+f_nonaggressive(x) = g_nonaggressive(x, 1)
+f_aggressive(x) = g_aggressive(x, 1)
+
+# The first test just makes sure that improvements to the compiler don't
+# render the annotation effectless.
+@test Base.return_types(f_nonaggressive, Tuple{Int})[1] == Val
+@test Base.return_types(f_aggressive, Tuple{Int})[1] == Val{1}
```
Comment (JeffBezanson) [inline test/core.jl:7522]: I'd move these to test/compiler/inference

File: stdlib/Serialization/src/Serialization.jl [hunk 4]
```diff
@@ -22,7 +22,8 @@ mutable struct Serializer{I<:IO} <: AbstractSerializer
     table::IdDict{Any,Any}
     pending_refs::Vector{Int}
     known_object_data::Dict{UInt64,Any}
-    Serializer{I}(io::I) where I<:IO = new(io, 0, IdDict(), Int[], Dict{UInt64,Any}())
+    version::Int
```
Comment (JeffBezanson) [inline stdlib/Serialization/src/Serialization.jl:25]: This is probably a good idea, but I still usually try to handle these by making the data fully self-describing, i.e. store the new field in such a way that the type of value you deserialize tells you which version of the object you have.

Comment (Keno) [reply stdlib/Serialization/src/Serialization.jl:25]: That's what I tried at first, but it just becomes super messy, for what seems to be little benefit. I don't see a realistic way to support non-versioned serialization blobs stably, so we might as well require the version.

**Comment by Keno** [comment]
I've rebased this. @JeffBezanson could I bother you to change the Serialization code to your satisfaction. I'm still not entirely convinced I understand why we can't just do the simple version-based thing - the current tricks in there just seem extremely brittle and non-scalable, but if you'd prefer the other thing, I guess that's ok - I just the feature ;).

**Comment by vtjnash** [comment]
Why is this useful, when it's not inlined, so this code will typically not be what ends up getting run? Do we need instead some degrees of `@inline` strength?

**Comment by Keno** [comment]
Why wouldn't it be useful? Aggressive constprop affects the return type we infer, so even if it's not inlined, sharper type information in the caller can make for a much more efficient function.

**Review by JeffBezanson — APPROVED** [review]

**Comment by JeffBezanson** [comment]
Ok, I think we might as well merge this as-is.

File: stdlib/Serialization/src/Serialization.jl [hunk 5]
```diff
@@ -716,6 +721,7 @@ function readheader(s::AbstractSerializer)
         error("""Cannot read stream serialized with a newer version of Julia.
                  Got data version $version > current version $ser_version""")
     end
+    s.version = version
 end
```
Comment (vtjnash) [inline stdlib/Serialization/src/Serialization.jl:725]:
```suggestion
    nothing
end
```

File: test/core.jl [hunk 6]
```diff
@@ -7527,4 +7527,4 @@ const RedefineVarargN{N} = Tuple{Vararg{RedefineVararg, N}}
 const RedefineVarargN{N} = Tuple{Vararg{RedefineVararg, N}}
 
 # NTuples with non-types
-@test NTuple{3, 2} == Tuple{2, 2, 2}
+@test NTuple{3, 2} == Tuple{2, 2, 2}
```
Comment (vtjnash) [inline test/core.jl:7530]: unrelated (and slightly worse)

**Comment by timholy** [comment]
Does it make any sense to also have a way of blocking constant-prop? I'm thinking #38983. It seems pretty cheap to reinfer for specific constants, so :man_shrugging:

**Comment by JeffBezanson** [comment]
Yes that would make sense. We've seen cases (though it's fairly rare) where it's useful to manually block constant prop.

**Comment by vtjnash** [comment]
We have `@noinline` already that will do that, but I think the observation there is that we want to consider `@nospecialize` and avoid inspecting specific typed specializations.

**Comment by timholy** [comment]
> We have @noinline already that will do that, but I think the observation there is that we want to consider @nospecialize and avoid inspecting specific typed specializations.

Good point although I'm pretty sure `@noinline` doesn't prevent constant-prop on `Core.kwfunc` and possibly (haven't checked) `Base.bodyfunction` methods.

**Comment by timholy** [comment]
I don't want to derail this from its original purpose, so please see https://github.com/JuliaLang/julia/issues/38983#issuecomment-757143042.

**Pull request merged**